src/Microsoft.AspNetCore.OData/Extensions/HttpRequestExtensions.cs

```diff
@@ -451,6 +454,56 @@ public static void DeleteRequestContainer(this HttpRequest request, bool dispose
             }
         }
 
+        /// <summary>
+        /// Checks whether the request is a POST targeted at a resource path ending in /$query.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        /// <param name="oDataPath">The OData path.</param>
+        /// <returns>true if the request path has $query segment.</returns>
+        internal static bool IsQueryRequest(this HttpRequest request, string oDataPath)
+        {
+            return request.Method.Equals(HttpMethods.Post) && 
+                oDataPath?.TrimEnd('/').EndsWith('/' + ODataRouteConstants.QuerySegment, StringComparison.OrdinalIgnoreCase) == true;
+        }
+
+        /// <summary>
+        /// Transforms a POST request targeted at a resource path ending in $query into a GET request. 
+        /// The query options are parsed from the request body and appended to the request URL.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        /// <param name="requestContainerFactory">Request container factory.</param>
+        internal static void TransformQueryRequest(this HttpRequest request, Func<IServiceProvider> requestContainerFactory)
+        {
+            if (requestContainerFactory == null)
+            {
+                throw Error.ArgumentNull("requestContainerFactory");
+            }
+
+            IServiceProvider requestContainer = requestContainerFactory();
+
+            // Fetch parsers available in the request container for parsing the query options in the request body
+            IEnumerable<IODataQueryOptionsParser> queryOptionsParsers = requestContainer.GetRequiredService<IEnumerable<IODataQueryOptionsParser>>();
+            IODataQueryOptionsParser queryOptionsParser = queryOptionsParsers.FirstOrDefault(d => d.CanParse(request));
+
+            if (queryOptionsParser == null)
+            {
+                throw new ODataException(string.Format(
+                    CultureInfo.InvariantCulture, 
+                    SRResources.CannotFindParserForRequestMediaType,
+                    request.ContentType));
+            }
+
+            string queryString = queryOptionsParser.ParseAsync(request.Body).Result;
+
+            // Request path starts with a /
+            string requestPath = request.Path.Value;
+            requestPath = requestPath.Substring(0, requestPath.LastIndexOf('/' + ODataRouteConstants.QuerySegment, StringComparison.OrdinalIgnoreCase));
+
+            request.Path = new PathString(requestPath);
+            request.QueryString = new QueryString(queryString);
```

Comment:
>QueryString(queryString) [](start = 38, length = 24)

as per above, the spec says that query options in the body can be combined with query options in the request.

Reply:
@mikepizzo Fixed

Reply:
Slightly cleaner (so you don't have to check IsNullOrWhitespace twice):

            if (!string.IsNullOrWhiteSpace(queryOptions))
            {
                   if(string.IsNullOrWhiteSpace(queryString)
                   {
                        queryString = '?' + queryOptions;
                   }
                   else
                  {
                        queryString += ('&' + queryOptions);
                  }
            }


---
In reply to: [488699478](https://github.com/OData/WebApi/pull/2191#discussion_r488699478) [](ancestors = 488699478)